lib/lf/fe/fe_space_hierarchic.h

```diff
@@ -0,0 +1,141 @@
+#include "hierarchic_fe.h"
+#include "scalar_fe_space.h"
+
+namespace lf::fe {
+
+/**
+ * @brief Lagrangian Finite Element Space of arbitrary degree
+ *
+ * @tparam SCALAR underlying scalar type, usually either `double` or
+ * `complex<double>`
+ *
+ * This FE Space contains hierarchic Finite Elements, meaning that
+ * the function spaces of lower polynomial degree are contained
+ * in the higher order function spaces.
+ *
+ * The polynomial degree is constant over the whole mesh meaning
+ * that there is no \f$p\f$-refinement supported.
+ *
+ * @note Some of the pointers may be NULL. For instance, if all computations
+ *       are done on purely triangular meshes then a finite element
+ * specification for quadrilaterals need not be given.
+ */
+template <typename SCALAR>
+class FeSpaceHierarchic : public ScalarFESpace<SCALAR> {
+ public:
+  using Scalar = SCALAR;
+
+  FeSpaceHierarchic() = delete;
+  FeSpaceHierarchic(const FeSpaceHierarchic &) = delete;
+  FeSpaceHierarchic(FeSpaceHierarchic &&) noexcept = default;
+  FeSpaceHierarchic &operator=(const FeSpaceHierarchic &) = delete;
+  FeSpaceHierarchic &operator=(FeSpaceHierarchic &&) noexcept = default;
+
+  /**
+   * @brief Constructor: Sets up the dof handler
+   * @param mesh_p A shared pointer to the underlying mesh (immutable)
+   * @param N The polynomial degree of the Finite Element Space
+   */
+  explicit FeSpaceHierarchic(
+      const std::shared_ptr<const lf::mesh::Mesh> &mesh_p, unsigned N)
+      : ScalarFESpace<SCALAR>(mesh_p), ref_el_(mesh_p, nullptr), degree_(N) {
+    init();
+  }
+
+  using ScalarFESpace<SCALAR>::Mesh;
+
+  /** @brief access to associated local-to-global map
+   * @return a reference to the lf::assemble::DofHandler object (immutable)
+   */
+  [[nodiscard]] const lf::assemble::DofHandler &LocGlobMap() const override {
+    LF_VERIFY_MSG(Mesh() != nullptr, "No valid FE space object: no mesh");
+    LF_VERIFY_MSG(dofh_p_ != nullptr,
+                  "No valid FE space object: no dof handler");
+    return *dofh_p_;
+  }
+
+  /** @brief access to shape function layout for cells
+   * @copydoc SclarFESpace::ShapeFunctionLayout(const lf::mesh::Entity&)
+   */
+  [[nodiscard]] std::shared_ptr<const ScalarReferenceFiniteElement<SCALAR>>
+  ShapeFunctionLayout(const lf::mesh::Entity &entity) const override {
+    return ref_el_(entity);
+  }
+
+  /** @brief number of _interior_ shape functions associated to entities of
+   * various types
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      const lf::mesh::Entity &entity) const override {
+    return ShapeFunctionLayout(entity)->NumRefShapeFunctions();
+  }
+
+  ~FeSpaceHierarchic() override = default;
+
+ private:
+  // R.H.: Should this really be shared pointers ?
+  lf::mesh::utils::AllCodimMeshDataSet<
+      std::shared_ptr<const lf::fe::ScalarReferenceFiniteElement<SCALAR>>>
+      ref_el_;
+  std::unique_ptr<lf::assemble::UniformFEDofHandler> dofh_p_;
```

Comment:
The problem is that `lf::assemble::UniformFEDofHandler` is not default-constructible. And because one needs quite a complex function in order to build its constructor arguments, it's not easy to initialize it in the member initializer list. The same problem occurs in the implementation of `lf::uscalfe::UniformScalarFESpace` and I took the solution with using a pointer which is first initialized to `nullptr` from there.

Reply:
Aha I see. I wonder a bit why `UniformScalarFESpace` / `ScalarFESpace` are required to be default constructible. There is a comment  in the code that says that this is because of `std::vector`, but std::vector can also be used with types that are not default constructible...?

Also I couldn't find any `std::vector<UniformScalarFESpace>` in all of LehrFEM++. And even if we would need an FESpace to be default constructible, we could still use `std::optional<UniformScalarFESpace>`.
Therefore I propose that we remove the default constructor of `UniformScalarFESpace` and `ScalarFESpace`.

Reply:
That wouldn't solve the problem though. The problem is that we cannot initialize the `lf::assemble::UniformFEDofHandler` in the member initializer list of the FE Spaces. To be able to do that, we would need to make `lf::assemble::UniformFEDofHandler` default constructible.

Reply:
Aha, but wait, this can be solved, I'll make a commit right now.

Reply:
Okay I've resolved this problem (and a few related ones) with https://github.com/craffael/lehrfempp/commit/9608889bb2aca8a1c4b76cdb26a526809c43c069